File: leetcode/273-integer-to-english-words/main.go
package main

import (
	"fmt"
	"strings"
)

/*
	1st approach:
	1. separate the numbers by 3 didgits, 123456789 -> 123, 456, 789
	2. for each 3 digits, translate to english
	3. for append Thousand, Million and Billion for each iteration of division

	Time		O(n)
	Space		O(n)
	0 ms, faster than 100.00%
	26feb2019
*/
func numberToWords(num int) string {
	d := []string{"", "Thousand", "Million", "Billion"}
	temp := ""
	i := 0
	for true {
		remain := num % 1000
		num = num / 1000
		threeDigitsWords := threeDigitsToWords(remain)
		if len(threeDigitsWords) > 0 {
			temp = threeDigitsWords + " " + d[i] + " " + temp
		}
		i++
		if num == 0 {
			break
		}
	}
	temp = strings.TrimSpace(temp)
	if len(temp) == 0 {
		return "Zero"
	}
	return temp
}

func threeDigitsToWords(num int) string {
	digits := []string{"", "One", "Two", "Three", "Four", "Five", "Six", "Seven", "Eight", "Nine", "Ten",
		"Eleven", "Twelve", "Thirteen", "Fourteen", "Fifteen", "Sixteen", "Seventeen", "Eighteen", "Nineteen"}
	tens := []string{"", "", "Twenty", "Thirty", "Forty", "Fifty", "Sixty", "Seventy", "Eighty", "Ninety"}
	a := num / 100
	num = num % 100
	temp := ""
	if a > 0 {
		temp += digits[a] + " Hundred "
	}
	if num < 20 {
		temp += digits[num]
	} else {
		b := num / 10
		num = num % 10
		temp += tens[b]
		temp += " " + digits[num]
	}
	return strings.TrimSpace(temp)
}

func main() {
	fmt.Println(threeDigitsToWords(999))
	fmt.Println(threeDigitsToWords(990))
	fmt.Println(threeDigitsToWords(900))
	fmt.Println(threeDigitsToWords(120))
	fmt.Println(threeDigitsToWords(102))
	fmt.Println(threeDigitsToWords(919))
	fmt.Println(threeDigitsToWords(99))
	fmt.Println(threeDigitsToWords(50))
	fmt.Println(threeDigitsToWords(20))
	fmt.Println(threeDigitsToWords(19))
	fmt.Println(threeDigitsToWords(10))
	fmt.Println(threeDigitsToWords(1))
	fmt.Println(threeDigitsToWords(0))

	fmt.Println(numberToWords(999))
	fmt.Println(numberToWords(990))
	fmt.Println(numberToWords(900))
	fmt.Println(numberToWords(120))
	fmt.Println(numberToWords(102))
	fmt.Println(numberToWords(919))
	fmt.Println(numberToWords(99))
	fmt.Println(numberToWords(50))
	fmt.Println(numberToWords(20))
	fmt.Println(numberToWords(19))
	fmt.Println(numberToWords(10))
	fmt.Println(numberToWords(1))
	fmt.Println(numberToWords(0))
	fmt.Println(numberToWords(1000))
	fmt.Println(numberToWords(9909))
	fmt.Println(numberToWords(9912))
	fmt.Println(numberToWords(9999))
	fmt.Println(numberToWords(45000))
	fmt.Println(numberToWords(40400))
	fmt.Println(numberToWords(12500))
	fmt.Println(numberToWords(12012))
	fmt.Println(numberToWords(1000000))
	fmt.Println(numberToWords(1234567))
	fmt.Println(numberToWords(1001000000))
	fmt.Println(numberToWords(1234567891))
}

File: leetcode/273-integer-to-english-words/main.py
"""
    1st approach:
	1. separate the number portion by portion with 3 digits, 123456789 -> 123, 456, 789
	2. for each 3 digits, translate to english
	3. append Thousand, Million and Billion in between the portions

	Time		O(N)
	Space		O(N)
	20 ms, faster than 100.00%
"""


class Solution(object):
    def numberToWords(self, num):
        """
        :type num: int
        :rtype: str
        """
        if num == 0:
            return "Zero"
        d = ["", "Thousand", "Million", "Billion"]
        result = ""
        i = 0
        while num > 0:
            remain = num % 1000
            num = num // 1000
            threeDigitWord = self.threeDigitsToWords(remain)
            # imagine if we have 1,000,000, we should check if the '3digits' is empty before appending to the result
            if len(threeDigitWord) > 0:
                result = threeDigitWord + " " + d[i] + " " + result
            i += 1
        return result.strip()

    def threeDigitsToWords(self, num):
        digits = ["", "One", "Two", "Three", "Four", "Five", "Six", "Seven", "Eight", "Nine", "Ten", "Eleven",
                  "Twelve", "Thirteen", "Fourteen", "Fifteen", "Sixteen", "Seventeen", "Eighteen", "Nineteen"]
        tens = ["", "", "Twenty", "Thirty", "Forty",
                "Fifty", "Sixty", "Seventy", "Eighty", "Ninety"]
        a = num // 100
        bc = num % 100
        temp = ""
        if a > 0:
            temp += digits[a] + " Hundred"
        if bc < 20:
            temp += " " + digits[bc]
        else:
            b = bc // 10
            c = bc % 10
            temp += " " + tens[b]
            temp += " " + digits[c]
        return temp.strip()


"""
    2nd: similar to 1st but using an array instead of a string for the result construction

    Time		O(N)
	Space		O(N)
    20 ms, faster than 70.44%
"""


class Solution(object):
    def numberToWords(self, num):
        """
        :type num: int
        :rtype: str
        """
        if num == 0:
            return "Zero"
        d = ["", "Thousand", "Million", "Billion"]
        result = []
        i = 0
        while num > 0:
            remain = num % 1000
            num = num // 1000
            threeDigitWord = self.threeDigitsToWords(remain)
            # imagine if we have 1,000,000, we should check if the '3digits' is empty before appending to the result
            if len(threeDigitWord) > 0:
                result.insert(0, threeDigitWord + ' ' + d[i])
            i += 1
        return ' '.join(result).strip()

    def threeDigitsToWords(self, num):
        digits = ["", "One", "Two", "Three", "Four", "Five", "Six", "Seven", "Eight", "Nine", "Ten", "Eleven",
                  "Twelve", "Thirteen", "Fourteen", "Fifteen", "Sixteen", "Seventeen", "Eighteen", "Nineteen"]
        tens = ["", "", "Twenty", "Thirty", "Forty",
                "Fifty", "Sixty", "Seventy", "Eighty", "Ninety"]
        a = num // 100
        bc = num % 100
        expression = []
        if a > 0:
            expression.append(digits[a])
            expression.append("Hundred")
        if bc < 20:
            expression.append(digits[bc])
        else:
            b = bc // 10
            c = bc % 10
            expression.append(tens[b])
            expression.append(digits[c])
        return ' '.join(expression).strip()
